server/asset/eth/tokener.go

```diff
@@ -0,0 +1,150 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+//go:build lgpl
+// +build lgpl
+
+package eth
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	"decred.org/dcrdex/dex"
+	"decred.org/dcrdex/dex/networks/erc20"
+	erc20v0 "decred.org/dcrdex/dex/networks/erc20/contracts/v0"
+	dexeth "decred.org/dcrdex/dex/networks/eth"
+	swapv0 "decred.org/dcrdex/dex/networks/eth/contracts/v0"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// swapContract is a generic source of swap contract data.
+type swapContract interface {
+	Swap(context.Context, [32]byte) (*dexeth.SwapState, error)
+}
+
+// tokenAddresser exposes the TokenAddress method of a token swap contract.
+type tokenAddresser interface {
+	TokenAddress(*bind.CallOpts) (common.Address, error)
+}
+
+// erc2Contract exposes methods of a token's ERC20 contract.
+type erc20Contract interface {
+	BalanceOf(*bind.CallOpts, common.Address) (*big.Int, error)
+}
+
+// tokener is a contract data manager for a token.
+type tokener struct {
+	*registeredToken
+	swapContract
+	tokenAddresser
+	erc20Contract
+	ver                     uint32
```

Comment:
To clarify, `tokener.ver` here is the version specific to this token, changed if the `erc20Contract` changed, such as if USDT got a new contract address.  Whereas `registeredToken.ver` is the version of the ERC20 swap contract, changed if `swapContract` (and thus `tokenAddresser`) changed, such as if we redeployed a new dex/networks/erc20/contracts/ERC20SwapVx.sol (for either the same erc20 address or a new one).  Is that how the fields and versions are related?

Reply:
All versions refer to the swap contract version and swap contract address, updated by us only if a new swap contract is implemented or if gases need to be updated for some reason. We have no mechanism to update the token erc20 contract address.

Reply:
Ok. So what I'm seeing is that `tokener` has a `ver` from the embedded `registereToken` and another one at this comment, which overrides/shadows the embedded one.  They are intended to be the same from what I understand now.

I think you can get rid of this explicit `tokener.ver` field then.